layer/layer_store.go

```diff
@@ -106,11 +106,17 @@ func NewStoreFromGraphDriver(store MetadataStore, driver graphdriver.Driver) (St
 	return ls, nil
 }
 
-func (ls *layerStore) loadLayer(layer ChainID) (*roLayer, error) {
+func (ls *layerStore) loadLayer(layer ChainID) (l *roLayer, err error) {
 	cl, ok := ls.layerMap[layer]
 	if ok {
 		return cl, nil
 	}
+	defer func() {
+		// If failed to load the layer, remove the layer metadata.
+		if err != nil {
+			ls.store.Remove(layer)
+		}
+	}()
```

Comment:
The main problem to keep the metadata on failures is we can't load/pull the same image any more, it will cause rename error something like 
```
rename /var/lib/docker/image/devicemapper/layerdb/tmp/layer-870197081 /var/lib/docker/image/devicemapper/layerdb/sha256/b257bb4b4aeaa5590b31c125581c301d69352345c09fe860a35de4c7cf212ae7: directory not empty
```
we have to remove the directory manually 

 @aaronlehmann  Instead of removing the metadata in a `defer`, how about remove it on failuer which is can't caused by temporay condition? And which failure will caused by temporary condition?